flink-ml-lib/src/main/java/org/apache/flink/ml/common/allreduce/AllReduceUtils.java

```diff
@@ -0,0 +1,314 @@
+                        .transform(
+                                "all-reduce-sum",
+                                new TupleTypeInfo<>(
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO),
+                                new AllReduceSum())
+                        .name("all-reduce-sum");
+
+        return allReduceSum
+                .partitionCustom((key, numPartitions) -> key % numPartitions, x -> x.f0)
+                .transform(
+                        "all-reduce-recv", TypeInformation.of(double[].class), new AllReduceRecv())
+                .name("all-reduce-recv");
+    }
+
+    /**
+     * Splits each double array into multiple pieces and send each piece to the corresponding
+     * partition.
+     */
+    private static class AllReduceSend
+            extends AbstractStreamOperator<Tuple4<Integer, Integer, Integer, double[]>>
+            implements OneInputStreamOperator<
+                            double[], Tuple4<Integer, Integer, Integer, double[]>>,
+                    BoundedOneInput {
+
+        double[] receivedElement;
+
+        double[] transBuf = new double[TRANSFER_BUFFER_SIZE];
+
+        @Override
+        public void endInput() {
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+            for (int taskId = 0; taskId < numTasks; taskId++) {
+                int startPieceId =
+                        DistributedInfo.getStartPieceId(taskId, numTasks, receivedElement.length);
+                int numPiecesToHandle =
+                        DistributedInfo.getNumPiecesByTaskId(
+                                taskId, numTasks, receivedElement.length);
+                for (int piece = 0; piece < numPiecesToHandle; piece++) {
+                    int bufStart = (startPieceId + piece) * TRANSFER_BUFFER_SIZE;
+                    System.arraycopy(
+                            receivedElement,
+                            bufStart,
+                            transBuf,
+                            0,
+                            DistributedInfo.getLengthOfPiece(
+                                    startPieceId + piece, receivedElement.length));
+                    output.collect(
+                            new StreamRecord<>(
+                                    Tuple4.of(
+                                            taskId,
+                                            startPieceId + piece,
+                                            receivedElement.length,
+                                            transBuf)));
+                }
+            }
+        }
+
+        @Override
+        public void processElement(StreamRecord<double[]> streamRecord) {
+            Preconditions.checkState(
```

Comment:
Yes.

Reply:
If it can be triggered by errors in user-inputs, then it should be IllegalArgumentException (could be something else) instead of IllegalStateException, right?

Reply:
Thanks for the explanation. I have changed to throw RuntimeException there.